File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.py
import sys
input = sys.stdin.readline

m,n = map(int, input().split())
lenArr = list(map(int, input().split()))

start = 1 #과자 길이가 1 이상이기 때문에 start = 1로 시작
end = int(1e9)

ans = 0
while start <= end:
    # mid: 이 과자길이로 나눠주면 어떨까에 대한 값
    mid = (end + start) // 2
    
    # c: mid사이즈로 모든 과자들을 나누었을 떄 나오는 과자 갯수
    c = 0
    for l in lenArr:
        c += l // mid

    # mid사이즈로 만들어진 과자갯수c가 조카들의 수m보다 많거나 같으면
    # 현재 답과 mid를 비교하여 더 큰 값을 답으로 정하고
    # mid 보다 더 큰 크기로 과자를 나누어줄 수 있는지 탐색한다.
    if c >= m:
        ans = max(ans, mid)
        start = mid+1
    # 만약 과자갯수c가 조카들의 수m보다 적으면
    # mid보다 작은 크기의 과자로 탐색한다
    else:
        end = mid-1
print(ans)

File: 백준/Silver/16401. 과자 나눠주기/과자 나눠주기.java

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;

public class Main {
	public static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

	public static void main(String args[]) throws IOException {
		Integer[] mn = input();
		int m = mn[0];
		int n = mn[1];
		int result = 0;

		Integer[] input = input();

		binarySearch(m, result, input);
	}

	private static void binarySearch(int m, int result, Integer[] input) {
		int start = 1,
			end = 1000000000;
		int mid;
		int cookieCount = 0;
		while (start <= end) {
			mid = (start + end) / 2;

			for(int i : input) {
				cookieCount += i/mid;
			}

			if(cookieCount >= m) {
				start = mid + 1;
				result = Math.max(mid, result);
			}
			else if(cookieCount < m) {
				end = mid - 1;
			}

			cookieCount = 0;
		}

		System.out.print(result);
	}

	private static Integer[] input() throws IOException {
		return Arrays.stream(br.readLine().split(" "))
			.map(Integer::parseInt)
			.toArray(Integer[]::new);
	}
}
